### ISSUE TITLE
Post install script fails with Expected "0.21.5" but got "0.21.4"

### ISSUE THREAD

**Issue by aminya** [evanw/esbuild#3800]
When I run `pnpm install`, I get the following error.
```js
│     throw new Error(`Expected ${JSON.stringify(versionFromPackageJSON)} but…
│           ^
│ Error: Expected "0.21.5" but got "0.21.4"
│     at validateBinaryVersion (./test
│     at ./test/node_modules/esbuild/…
│ Node.js v20.14.0
```

Esbuild is my indirect dependency, and I don't have control over the version, and in the lock file, the versions are correct:
```

  '@esbuild/linux-x64@0.21.5':
    resolution: {integrity: sha512-1rYdTpyv03iycF1+BhzrzQJCdOuAOtaqHTWJZCWvijKD2N5Xu0TtVC8/+1faWqcP9iBCWOmjmhoH94dH82BxPQ==}
    engines: {node: '>=12'}
    cpu: [x64]
    os: [linux]

  esbuild@0.21.5:
    resolution: {integrity: sha512-mg3OPMV4hXywwpoDxu3Qda5xCKQi+vCTZq8S9J/EpkhB2HzKXq4SNFZE3+NK93JYxc8VMSep+lOUSC/RVKaBqw==}
    engines: {node: '>=12'}
    hasBin: true

```

```
pnpm --version
9.3.0
```

**Comment by evanw** [comment]
Nothing has changed about esbuild’s installer recently. There was also [another recent report](https://github.com/evanw/esbuild/issues/3796) of pnpm’s installer misbehaving. I’m guessing that pnpm has introduced a regression that is affecting esbuild. I suggest you file a bug with them and either use an older version of pnpm or use npm in the meantime. It would also be good to know which versions of pnpm have this bug and which do not.

**Comment by aminya** [comment]
I suspect it is related to using the hoisted node-linker and the fact that it is an indirect dependency. I've reported it in the pnpm repo

```
node-linker=hoisted
package-manager-strict=false
```

**Comment by evanw** [comment]
I'm closing this as this appears to be a problem with the package manager, not with the esbuild package.

**Comment by httpete** [comment]
Hitting this here as well

```
npm error path /blahblah/react-components/node_modules/esbuild
npm error command failed
npm error command sh -c node install.js
npm error /Users/blahblah/builds/PC/Portal/console/website/src/main/webapp/react-components/node_modules/esbuild/install.js:133
npm error     throw new Error(Expected ${JSON.stringify(versionFromPackageJSON)} but got ${JSON.stringify(stdout)});
npm error           ^
npm error
npm error Error: Expected "0.20.2" but got "0.21.5"
```

**Comment by nathanbirrell** [comment]
Also getting this issue in CI. 

npm 10.7.0
pnpm 8.15.8
Node v18.20.4

**Comment by jodyhoonstarr** [comment]
Another workaround. Remove node_modules and the pnpm-lock file and try again. Below the alias `pnpm clean:node_modules` deletes all node folders in a monorepo. Ymmv.

```
❯ pnpm add esbuild@latest -w -D
│     throw new Error(`Expected ${JSON.stringify(versionFromPackageJSON)} but got ${JSON.stringify(stdout)}`);
│           ^
│ Error: Expected "0.23.1" but got "0.23.0"

❯ rm pnpm-lock.yaml
❯ pnpm clean:node_modules

❯ pnpm add esbuild@latest -w -D
node_modules/.pnpm/esbuild@0.23.1/node_modules/esbuild: Running postinstall script, done in 105ms
```

**Comment by darrylsepeda** [comment]
Currently stumbled on this issue, removing node_modules and pnpm-lock doesn't work for me.
esbuild is neither mentioned as dependencies nor devDependencies in my project.
adding/removing esbuild is also not a solution for me.

![Uploading image.png…]()

**Comment by aaronklinker-st** [comment]
I setup an `overrides` in my `package.json` and set it to the newer version of the two... Dangerous, but does the trick.

```json
{
  "pnpm": {
    "overrides": {
      "esbuild": "0.23.1"
    }
  }
}
```

**Comment by vizonluo** [comment]
I resolved that by upgrade node.js from v16 to v18

**Comment by HiKWang** [comment]
> I resolved that by upgrade node.js from v16 to v18

It works for me. Thanks.

**Comment by HiKWang** [comment]
> I resolved that by upgrade node.js from v16 to v18

It works for me. Thanks @vizonluo

**Comment by ReDrUm** [comment]
I'm getting stung by this also using Yarn 4 with `nodeLinker: node-modules`:

``` shell
Error: Expected "0.23.1" but got "0.23.0"
    at validateBinaryVersion (/path/to/project/node_modules/tsx/node_modules/esbuild/install.js:135:11)
    at /path/to/project/node_modules/tsx/node_modules/esbuild/install.js:283:5
```

Perhaps we could change the version enforcemnt from strict equality of the full semver version, to instead only consider the major/minor? e.g.

``` diff
+ const resolvedMinor = stdout.substring(0, stdout.lastIndexOf('.'));
+ const expectedMinor = stdout.substring(0, versionFromPackageJSON.lastIndexOf('.'));

- if (stdout !== versionFromPackageJSON) {
+ if (resolvedMinor !== expectedMinor) {
-    throw new Error(`Expected ${JSON.stringify(versionFromPackageJSON)} but got ${JSON.stringify(stdout)}`);
+    throw new Error(`Expected ${expectedMinor} but got ${resolvedMinor}`);
}
```

**Comment by aminya** [comment]
Okay, so the issue might be that esbuild doesn't pin the versions resulting in package managers installing newer versions.

**Comment by lomholdt** [comment]
> Another workaround. Remove node_modules and the pnpm-lock file and try again. Below the alias `pnpm clean:node_modules` deletes all node folders in a monorepo. Ymmv.
> 
> ```
> ❯ pnpm add esbuild@latest -w -D
> │     throw new Error(`Expected ${JSON.stringify(versionFromPackageJSON)} but got ${JSON.stringify(stdout)}`);
> │           ^
> │ Error: Expected "0.23.1" but got "0.23.0"
> 
> ❯ rm pnpm-lock.yaml
> ❯ pnpm clean:node_modules
> 
> ❯ pnpm add esbuild@latest -w -D
> node_modules/.pnpm/esbuild@0.23.1/node_modules/esbuild: Running postinstall script, done in 105ms
> ```

Thanks! I had this issue using `bun` but same error. Fixed it by deleting `bun.lockb` file and `node_modules` and re-ran `bun install`.

**Comment by alexrecuenco** [comment]
This is also happening to us without pnpm, just npm, when we updated vitest and vite to the latest version

**Comment by conorzhong** [comment]
For me, using `npm i -g pnpm` instead of `curl -fsSL https://get.pnpm.io/install.sh | sh -` solved the problem, hope this helps

**Comment by luiz00martins** [comment]
The only working fix for me was to nuke the directory, re-clone it, and re-install everything.

**Comment by richardwf1arcade** [comment]
> I resolved that by upgrade node.js from v16 to v18

v18 here. Doesn't work for me. Interestingly I've got it but colleagues don't.

**Comment by Alm0g** [comment]
I used Claude to help me summarize the solution, don't judge me :>
---
I had a version mismatch between the expected binary version (0.25.9), and the actual binary (0.25.5).
I had an outdated binary from a previous installation. Things like `pnpm store prune` wouldn't do...

I checked the postinstall script (`install.js`) which is where the errors originated from.
It calls `validateBinaryVersion()` which executes the binary, and checks if its `--version` output matches the package.json version.
The script expects `0.25.9` but gets `0.25.5` from the outdated binary, causing the validation to fail...

For me, **The solution** was to delete the old binary at `node_modules/.pnpm/esbuild@0.25.9/node_modules/esbuild/bin/esbuild` and rerun `pnpm install`.

```
Fix:
rm node_modules/.pnpm/esbuild@0.25.9/node_modules/esbuild/bin/esbuild
pnpm install
```
